File: packages/core/src/experiment/targeting.test.ts
import { describe, it, expect } from "vitest";
import { evaluateTargeting } from "./targeting.js";
import type { Factor } from "../types.js";
import type { TargetingRule, DeviceMetadata } from "./targeting.js";

function buildFactor(name: string, value: number): Factor {
  return {
    user_id: "user-1",
    component_path: "test/component",
    factor_name: name,
    factor_tier: "alarm",
    value,
    computed_at: new Date().toISOString(),
  };
}

describe("evaluateTargeting", () => {
  const factors: Factor[] = [
    buildFactor("error_rate", 0.25),
    buildFactor("rage_click_count", 5),
    buildFactor("completion_rate", 0.8),
  ];

  it("returns true when no rules are defined", () => {
    expect(evaluateTargeting(factors, [])).toBe(true);
  });

  // --- Legacy factor rules (backward compat, no type field) ---

  it("matches gt operator", () => {
    const rules: TargetingRule[] = [
      { factor: "error_rate", operator: "gt", threshold: 0.1 },
    ];
    expect(evaluateTargeting(factors, rules)).toBe(true);
  });

  it("rejects gt when value equals threshold", () => {
    const rules: TargetingRule[] = [
      { factor: "error_rate", operator: "gt", threshold: 0.25 },
    ];
    expect(evaluateTargeting(factors, rules)).toBe(false);
  });

  it("matches gte when value equals threshold", () => {
    const rules: TargetingRule[] = [
      { factor: "error_rate", operator: "gte", threshold: 0.25 },
    ];
    expect(evaluateTargeting(factors, rules)).toBe(true);
  });

  it("matches lt operator", () => {
    const rules: TargetingRule[] = [
      { factor: "completion_rate", operator: "lt", threshold: 0.9 },
    ];
    expect(evaluateTargeting(factors, rules)).toBe(true);
  });

  it("matches lte when value equals threshold", () => {
    const rules: TargetingRule[] = [
      { factor: "completion_rate", operator: "lte", threshold: 0.8 },
    ];
    expect(evaluateTargeting(factors, rules)).toBe(true);
  });

  it("matches eq operator", () => {
    const rules: TargetingRule[] = [
      { factor: "rage_click_count", operator: "eq", threshold: 5 },
    ];
    expect(evaluateTargeting(factors, rules)).toBe(true);
  });

  it("requires all rules to match (AND logic)", () => {
    const rules: TargetingRule[] = [
      { factor: "error_rate", operator: "gt", threshold: 0.1 },
      { factor: "completion_rate", operator: "lt", threshold: 0.5 },
    ];
    expect(evaluateTargeting(factors, rules)).toBe(false);
  });

  it("returns false when referenced factor is missing", () => {
    const rules: TargetingRule[] = [
      { factor: "nonexistent_factor", operator: "gt", threshold: 0 },
    ];
    expect(evaluateTargeting(factors, rules)).toBe(false);
  });

  // --- Typed factor rules ---

  it("matches typed factor rule with type='factor'", () => {
    const rules: TargetingRule[] = [
      { type: "factor", factor: "error_rate", operator: "gt", threshold: 0.1 },
    ];
    expect(evaluateTargeting(factors, rules)).toBe(true);
  });

  // --- Metadata targeting rules ---

  const device: DeviceMetadata = {
    os_name: "android",
    os_version: "14",
    manufacturer: "Samsung",
    model: "Galaxy S24",
    app_version: "2.1.0",
    platform: "android",
  };

  it("matches metadata eq (case-insensitive)", () => {
    const rules: TargetingRule[] = [
      { type: "metadata", field: "os_name", operator: "eq", value: "Android" },
    ];
    expect(evaluateTargeting([], rules, device)).toBe(true);
  });

  it("rejects metadata eq when value differs", () => {
    const rules: TargetingRule[] = [
      { type: "metadata", field: "os_name", operator: "eq", value: "ios" },
    ];
    expect(evaluateTargeting([], rules, device)).toBe(false);
  });

  it("matches metadata neq", () => {
    const rules: TargetingRule[] = [
      { type: "metadata", field: "os_name", operator: "neq", value: "ios" },
    ];
    expect(evaluateTargeting([], rules, device)).toBe(true);
  });

  it("matches metadata contains", () => {
    const rules: TargetingRule[] = [
      { type: "metadata", field: "model", operator: "contains", value: "Galaxy" },
    ];
    expect(evaluateTargeting([], rules, device)).toBe(true);
  });

  it("matches metadata gte for version comparison", () => {
    const rules: TargetingRule[] = [
      { type: "metadata", field: "os_version", operator: "gte", value: "14" },
    ];
    expect(evaluateTargeting([], rules, device)).toBe(true);
  });

  it("rejects metadata gte when version is lower", () => {
    const rules: TargetingRule[] = [
      { type: "metadata", field: "os_version", operator: "gte", value: "15" },
    ];
    expect(evaluateTargeting([], rules, device)).toBe(false);
  });

  it("returns false when metadata field is missing from device", () => {
    const rules: TargetingRule[] = [
      { type: "metadata", field: "app_build", operator: "eq", value: "100" },
    ];
    expect(evaluateTargeting([], rules, device)).toBe(false);
  });

  it("returns false for metadata rule when no device metadata provided", () => {
    const rules: TargetingRule[] = [
      { type: "metadata", field: "os_name", operator: "eq", value: "android" },
    ];
    expect(evaluateTargeting([], rules)).toBe(false);
  });

  // --- Mixed rules (factor + metadata) ---

  it("combines factor and metadata rules with AND logic", () => {
    const rules: TargetingRule[] = [
      { factor: "error_rate", operator: "gt", threshold: 0.1 },
      { type: "metadata", field: "os_name", operator: "eq", value: "android" },
    ];
    expect(evaluateTargeting(factors, rules, device)).toBe(true);
  });

  it("rejects when factor matches but metadata does not", () => {
    const rules: TargetingRule[] = [
      { factor: "error_rate", operator: "gt", threshold: 0.1 },
      { type: "metadata", field: "os_name", operator: "eq", value: "ios" },
    ];
    expect(evaluateTargeting(factors, rules, device)).toBe(false);
  });
});

File: packages/kotlin-engine/src/commonTest/kotlin/ai/factoredui/engine/experiments/TargetingTest.kt
package ai.factoredui.engine.experiments

import ai.factoredui.engine.factors.Factor
import ai.factoredui.engine.factors.FactorTier
import kotlin.test.Test
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class TargetingTest {

    private fun factor(name: String, value: Double) =
        Factor("u1", "checkout", name, FactorTier.ALARM, value, "2026-01-01T00:00:00Z")

    private val factors = listOf(factor("error_rate", 0.5), factor("completion", 0.8))

    @Test
    fun emptyRulesAlwaysEligible() {
        assertTrue(evaluateTargeting(factors, emptyList()))
    }

    @Test
    fun factorRuleComparesValue() {
        assertTrue(evaluateTargeting(factors, listOf(FactorRule("error_rate", TargetingOperator.GT, 0.3))))
        assertFalse(evaluateTargeting(factors, listOf(FactorRule("error_rate", TargetingOperator.GT, 0.6))))
        assertTrue(evaluateTargeting(factors, listOf(FactorRule("error_rate", TargetingOperator.EQ, 0.5))))
    }

    @Test
    fun missingFactorFailsRule() {
        assertFalse(evaluateTargeting(factors, listOf(FactorRule("nonexistent", TargetingOperator.GT, 0.0))))
    }

    @Test
    fun allRulesMustPass() {
        val rules = listOf(
            FactorRule("error_rate", TargetingOperator.GT, 0.3),
            FactorRule("completion", TargetingOperator.LT, 0.5),
        )
        assertFalse(evaluateTargeting(factors, rules)) // completion 0.8 is not < 0.5
    }

    @Test
    fun metadataRuleIsCaseInsensitive() {
        val metadata = DeviceMetadata(platform = "iOS", manufacturer = "Samsung", osVersion = "14")
        assertTrue(
            evaluateTargeting(
                emptyList(),
                listOf(MetadataRule(MetadataField.PLATFORM, MetadataOperator.EQ, "ios")),
                metadata,
            ),
        )
        assertTrue(
            evaluateTargeting(
                emptyList(),
                listOf(MetadataRule(MetadataField.MANUFACTURER, MetadataOperator.CONTAINS, "sams")),
                metadata,
            ),
        )
        assertTrue(
            evaluateTargeting(
                emptyList(),
                listOf(MetadataRule(MetadataField.OS_VERSION, MetadataOperator.GTE, "14")),
                metadata,
            ),
        )
    }

    @Test
    fun metadataRuleFailsWithoutMetadata() {
        assertFalse(
            evaluateTargeting(
                emptyList(),
                listOf(MetadataRule(MetadataField.PLATFORM, MetadataOperator.EQ, "ios")),
                deviceMetadata = null,
            ),
        )
    }

    @Test
    fun metadataRuleFailsWhenFieldAbsent() {
        assertFalse(
            evaluateTargeting(
                emptyList(),
                listOf(MetadataRule(MetadataField.MODEL, MetadataOperator.EQ, "pixel")),
                DeviceMetadata(platform = "android"),
            ),
        )
    }
}
